improver/wxcode/modal_code.py

```diff
@@ -244,28 +315,465 @@ def process(self, cubes: CubeList) -> Cube:
+        """
+        # Find indices corresponding to dry weather codes inclusive of a wet bias.
+        dry_counts = np.sum(
+            np.isin(cube.data, self.broad_categories["dry"]), axis=time_axis
+        )
+        wet_counts = np.sum(
+            np.isin(cube.data, self.broad_categories["wet"]), axis=time_axis
+        )
+        return dry_counts > self.wet_bias * wet_counts
+
+    @staticmethod
+    def counts_per_category(data: np.ndarray) -> np.ndarray:
+        """Implemented following https://stackoverflow.com/questions/46256279/bin-elements-per-row-vectorized-2d-bincount-for-numpy/46256361#46256361  # noqa: E501
+        Use np.bincount to count the number of occurrences within each category, so that
+        the most common occurrence can then be found.
+
+        Args:
+            data: Array where occurrences of each possible integer value between 0
+                and data.max() will be counted.
+
+        Returns:
+            An array of counts for the occurrence of each category within each row.
+        """
+        n_cat = data.max() + 1
+        a_offs = data + np.arange(data.shape[0])[:, None] * n_cat
+        # Use compressed() to avoid counting masked values.
+        return np.bincount(
+            a_offs.ravel().compressed(), minlength=data.shape[0] * n_cat
+        ).reshape(-1, n_cat)
+
+    def _find_most_significant_dry_code(
+        self, cube: Cube, result: Cube, dry_indices: np.ndarray, time_axis: int
+    ) -> Cube:
+        """Find the most significant dry weather code at each point.
+
+        Args:
+            cube: Weather code cube.
+            result: Cube into which to put the result.
+            dry_indices: Boolean, which is true if the weather codes at that point,
+                are dry.
+            time_axis: The time coordinate dimension.
+
+        Returns:
+            Cube where points that are dry are filled with the most common dry
+            code present at that point. If there is a tie, the most significant dry
+            weather code is used, assuming higher values for the weather code indicate
+            more significant weather.
+        """
+        data = cube.data.copy()
+        data = np.ma.masked_where(
+            ~np.isin(cube.data.copy(), self.broad_categories["dry"]), data
+        )
+
+        reshaped_data = data.T.reshape(-1, len(cube.coord("time").points), order="F")
+        counts = self.counts_per_category(reshaped_data)
+        # Add 1 to the final column, so that the highest index dry weather code is
+        # preferenced. This only has an effect in the event of ties. If adding 1
+        # creates a tie artifically, np.argmax will still automatically choose the
+        # lowest index result.
+        counts[:, -1] += 1
+        reshaped_counts = counts.reshape(*data.shape[1:], -1)
+        result.data[dry_indices] = np.argmax(reshaped_counts, axis=-1)[dry_indices]
+        return result
```

Comment:
This method doesn't currently hedge to the higher indices by default, which I thought was our preferred approach. Is this an active choice or a bug / oversight?

For this scenario: `[0, 2, 2, 0, 2, 1, 1, 1, 3, 3, 7, 8, 7, 7, 7, 7, 3, 3, 3, 1, 1, 0, 0, 0]`
We end up with a sunny symbol, despite this being tied with partly sunny:

![image](https://github.com/user-attachments/assets/c5e6285a-fa48-4e3e-9586-cf4869badecb)

There doesn't seem to be any flipping, though I am confused by all the reshaping, so possibly I've not understood correctly?

Reply:
I've adding flipping of the counts array, so that the weather codes with a higher index are preferentially selected by argmax.